商品購入機能

YYprogramer opened the pull request:
#what
商品購入機能の実装

#why
商品を購入できるようにするため

□テスト
○必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/8b191e3852cc8204a26b777ab9fdba53

○入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/172338a0f0bdb98ec00c52152c3f56b2

○ ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/2f0f45424b6c71d93e88c9197f8a4472

○ ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/92706abac5a18cad04465c4ae9dce389

○ ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/7b2aac0c9c32f32b907fd237fe9e0d97

○ 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/8b191e3852cc8204a26b777ab9fdba53

○ 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/fdac49d4d219862ae0fb70b5920d09e0

○ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/92706abac5a18cad04465c4ae9dce389

○ ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/81d738d140228e7288c36d944c12df08

○ テスト結果の画像
orderモデル　https://gyazo.com/d6693c2d44cd371506c666598009eba9
shipping_addressモデル　https://gyazo.com/0f7c82214fedd3ab6361e6f28d64d248

YYprogramer committed:
ordersコントローラーの作成

YYprogramer committed:
orderモデルの作成

YYprogramer committed:
sipping_addressモデルの作成

YYprogramer committed:
READMEの修正

・"purchase"から"order"に変更

YYprogramer committed:
ordersのルーティング

YYprogramer committed:
orders_controllerの作成

YYprogramer committed:
orderモデルの作成

YYprogramer committed:
order_formの作成

YYprogramer committed:
購入画面へのパスの設定

YYprogramer committed:
orderモデルのマイグレーションファイルの設定

YYprogramer committed:
SippingAddressのマイグレーションファイル作成

YYprogramer committed:
SippingAddressモデルのアソシエーション設定

YYprogramer committed:
Orderモデルのアソシエーション修正

has_one :shipping_addressとし
１対１のアソシエーションを設定

YYprogramer committed:
shippinng_addressモデルのタイピングミス

sipping_address ×
shipping_address ⚪︎

YYprogramer committed:
orderモデルのバリデーション設定

YYprogramer committed:
OrderFormModelの実装

YYprogramer committed:
orders_controller.rbの実装

YYprogramer committed:
ルーティングの記述蜜を修正

YYprogramer committed:
shipping_addressのマイグレート

YYprogramer committed:
購入画面ビューファイルの修正

YYprogramer committed:
order#indexへのルーティング

YYprogramer committed:
Orderコントローラのタイピングミスの修正

YYprogramer committed:
order#indexファイルの修正

YYprogramer committed:
routesのタイポ修正

YYprogramer committed:
ファイル名のタイポ修正

× oreder　
⚪︎ order

YYprogramer committed:
タイポ修正

YYprogramer committed:
購入画面のビュー実装

YYprogramer committed:
郵便番号のバリデーション

3桁の数字-4桁の数字
となっていないと保存できない

YYprogramer committed:
電話番号のバリデーション

電話番号は半角数字10桁〜11桁の場合のみ保存させる

YYprogramer committed:
PAYJPの実装

YYprogramer committed:
Card.jsの修正

タイポを修正
月と年を別々に取得するように修正

YYprogramer committed:
購入画面のビューファイル修正

hogeを適正な表記に修正

YYprogramer committed:
商品購入機能の実装

商品購入情報を各テーブルに保存できるように修正

YYprogramer committed:
商品購入機能の実装

各モデルに保存できるように修正

YYprogramer committed:
PayJP実装

クレジットカード情報をトークン化して発行するための実装

YYprogramer committed:
バリデーションの記述場所変更

YYprogramer committed:
Orderモデルバリデーション修正

YYprogramer committed:
バリデーションの削除

Formオブジェクト内でバリデーションを設定しているため、各モデル内でのバリデーションは削除しています

YYprogramer committed:
ShippingAddressModelへのデータ送信

YYprogramer committed:
OrderFormオブジェクトの実装

現在、商品購入ボタンを押した時に
・PayJPを介したTokenの発行
・Oredrsテーブルへ購入者及び商品情報の保存
は問題なく実行できるが
・購入者の住所情報の保存
ができない
このエラーを解消中

YYprogramer committed:
ユーザーのログイン状態によって遷移するページの処理

・自身が出品した商品の購入ページへ遷移しようとした場合
→トップページへ遷移
・未ログインユーザーが商品の購入ページへ遷移しようとした場合
→ログインページへ遷移

YYprogramer committed:
ShippingAddressのアクティブハッシュModule取り込み

YYprogramer committed:
ShippingAddressのテーブルからItemIDを削除

YYprogramer committed:
購入者情報とオーダー情報を関連付けする

YYprogramer committed:
商品一覧ページでSoldOut機能の実装

YYprogramer committed:
商品詳細機能の修正

売却済みの場合はSoldOutを表示させる

YYprogramer committed:
売却状況によって商品編集・商品削除機能ボタンの表示を変える

売却済みの場合は
自身が出品の商品であっても
商品編集・削除ボタンを表示させない

YYprogramer committed:
売却済み商品の編集機能について

売却済みの商品を編集しようとした場合
トップページに遷移する

YYprogramer committed:
新着商品は３つしか表示させない

YYprogramer committed:
OrderModelの単体テスト

YYprogramer committed:
フォームオブジェクトを使用したテストコードの修正

フォームオブジェクトを使用したテストコードを実施する場合は、各モデルごとではなくフォームオブジェクトでまとめて行う

YYprogramer committed:
Fakerを使用したテストコードの実装

YYprogramer committed:
OrderFormモデルの作成

YYprogramer committed:
order_formのテストコード実装

YYprogramer committed:
OrderFormのテストコード実施

CHURO-MAU commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,56 @@
+  before_action :redirect_unauthenticated_user,:redirect_own_item, only: :index
+
+  def index
+    @item = Item.find(params[:item_id])
```
５行目
こちらの記述は他のアクションでも、同様のものがございます。
private以下に新しくメソッドを作成し、before_actionで呼び出すようにしてみましょう。

YYprogramer replied:
了解しました
プライベートメソッド内で
set_itemメソッドとして取りまとめ
BeforeActionで呼び出しています

CHURO-MAU commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,56 @@
+  def create
+    @item = Item.find(params[:item_id])
+    @order_form = OrderForm.new(order_params)
+    @order_form.user_id = current_user.id
+    @order_form.item_id = @item.id
+    @order_form.token = params[:token]
```
１５〜１７行目
これらのカラム情報は、ストロングパラメーターでマージさせるように修正しておきましょう。

YYprogramer replied:
ストロングパラメータでマージさせました

CHURO-MAU commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,56 @@
+
+    if @order_form.valid?
+      pay_item
+      order = @order_form.save(user_id: current_user.id, item_id: @item.id)
```
２１行目
user_idやitem_idを引数として渡す必要はございませんので、@order_form.saveに修正しておきましょう。

YYprogramer replied:
削除しました

CHURO-MAU commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,31 @@
+  validates :token, presence: true
+
+
+  def save(user_id:,item_id:)
+    if valid?
+        order = Order.create(user_id: user_id, item_id: item_id)
+        shipping_address = ShippingAddress.create(
+          postal_code: postal_code, 
+          send_origin_id: send_origin_id,
+          city: city, 
+          street_address: street_address, 
+          building: building, 
+          phone_number: phone_number, 
+          order_id: order.id)
+      true
+    else
+      false
+    end
+  end
```
(user_id:,item_id:)および、条件分岐になりますので削除しておきましょう。

YYprogramer replied:
削除しました

CHURO-MAU commented on app/models/shipping_address.rb:
```diff
@@ -0,0 +1,12 @@
+  belongs_to :order
+  belongs_to :send_origin
+
+  validates :postal_code, presence: true
+  validates :send_origin_id, presence: true
+  validates :city, presence: true
+  validates :street_address, presence: true
+  validates :phone_number, presence: true
+  validates :order_id, presence: true
```
これらバリデーションは、Formオブジェクトに既に設定しているため削除しておきましょう。

YYprogramer replied:
削除しました

CHURO-MAU commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,31 @@
+  include ActiveModel::Model
+  attr_accessor :postal_code, :send_origin_id , :city, :street_address, :building, :phone_number, :user_id, :item_id, :token
+
+  validates :postal_code, format: { with: /\A\d{3}-\d{4}\z/, message: "Please enter in the format of '3-digit hyphen 4-digit'" }, presence: true
```
「with_options」を使用し、共通しているバリデーションをグループ化してみましょう。
これにより、コードがより簡潔にまとまります。

YYprogramer replied:
with_optionsを使用し
可読性を高めました

CHURO-MAU commented on app/views/items/show.html.erb:
```diff
@@ -27,12 +25,14 @@
+        <% end %>
       <% else %>
         <%# 商品が売れていない場合はこちらを表示しましょう %>
-        <%= link_to "購入画面に進む", "#" ,class:"item-red-btn"%>
+        <%= link_to "購入画面に進む", item_orders_path(@item) ,class:"item-red-btn"%>
         <%# //商品が売れていない場合はこちらを表示しましょう %>
```
こちら実装案内に関するコメントになりますので、削除しておきましょう。

もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめいたします。

YYprogramer replied:
削除しました

CHURO-MAU commented on app/models/item.rb:
```diff
@@ -7,8 +7,13 @@ class Item < ApplicationRecord
   belongs_to :send_time
   belongs_to :user
   has_one_attached :image
+  has_many :orders
```
１０行目
アソシエーションに誤りがございますので、見直してみましょう。

YYprogramer replied:
ordersとのアソシエーションを
has_manyからhas_oneに修正しました

CHURO-MAU commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+    end
+
+    context '商品購入が失敗する場合' do
```
異常系テストでは、これまでのテスト方法と同様に期待するエラーメッセージを確認するテストを行いますしょう。
```
 expect(@XXX.errors.full_messages).to include("〜〜〜")
```

YYprogramer replied:
エラーメッセージを確認するように修正しました

CHURO-MAU commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,84 @@
+require 'rails_helper'
+
+RSpec.describe OrderForm, type: :model do
```
[Orderモデルで行っていただいた異常系テスト](https://gyazo.com/d6693c2d44cd371506c666598009eba9)は、こちらのテストファイルでまとめて行うようにしましょう。

既にFormオブジェクトでも以下のバリデーションを設定いただいていますので、そちらに対するテストを記述いただければと思います。
```.rb
# app/models/order_form.rb
  validates :user_id, presence: true
  validates :item_id, presence: true
```

YYprogramer replied:
user_id及びitem_idが空白の場合保存できないという異常系のテストを行いました

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！

コメントいたしましたので、ご確認お願いいたします。
修正後は、再度レビュー依頼フォームのご提出をお願いいたします。

YYprogramer committed:
OrdersControllerのコード短縮化

@item = Item.find(params[:item_id])
をプライベートメソッド内で
set_itemメソッドとしてまとめ
BeforeActionで呼び出し、Index・Createアクション時に呼び出すようにしています

YYprogramer committed:
ストロングパラメータにマージする

    @order_form.user_id = current_user.id
    @order_form.item_id = @item.id
    @order_form.token = params[:token]
をストロングパラメータにマージさせる

YYprogramer committed:
引数の削除

YYprogramer committed:
ShippingAddressモデルの不要なバリデーションを削除

YYprogramer committed:
不要なコメントの削除

YYprogramer committed:
引数の削除及び条件分岐の削除

YYprogramer committed:
with＿optionsを使用しコードの可読性を高める

YYprogramer committed:
アソシエーションの修正

YYprogramer committed:
異常系テストの追加

YYprogramer committed:
異常系テストコードの修正

YYprogramer commented:
#what
商品購入機能の実装

#why
コードレビューの修正

□テスト
○ テスト結果の画像
OrderFormモデル　https://gyazo.com/78ec1f8043584828cfa0ed9abcdf52d2

maimai0401 commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,30 @@
+    validates :item_id
+    validates :token
+  end
+
```
ファイル差分が見つからなかったため、こちらにコメント失礼致します。
userモデルに購入モデルとのアソシエーションを追記しましょう。

YYprogramer replied:
UserModelにOrderModelとのアソシエーションを記述しました

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
コメントいたしましたので、ご確認お願いいたします。
修正後、再度レビュー依頼フォームにてご提出くださいませ。

YYprogramer committed:
UserModelとOrderModelのアソシエーション

YYprogramer commented:
#what
商品購入機能の実装

#why
コードレビューの修正

□修正内容
UserModelにOrderModelとのアソシエーションを追記

AsukaToyoshima commented on app/views/items/show.html.erb:
```diff
@@ -27,13 +25,13 @@
-        <p class="or-text">or</p>
-        <%= link_to "削除", item_path(@item), method: :delete, class:"item-destroy" %>
+        <% if !@item.sold_out? %>
+          <%= link_to "商品の編集", edit_item_path(@item), method: :get, class: "item-red-btn" %>
+          <p class="or-text">or</p>
+          <%= link_to "削除", item_path(@item), method: :delete, class: "item-destroy" %>
+        <% end %>
       <% else %>
-        <%# 商品が売れていない場合はこちらを表示しましょう %>
-        <%= link_to "購入画面に進む", "#" ,class:"item-red-btn"%>
-        <%# //商品が売れていない場合はこちらを表示しましょう %>
+        <%= link_to "購入画面に進む", item_orders_path(@item) ,class:"item-red-btn"%>
       <% end %>
     <% end %>
```
こちらの条件分岐ですと、売り切れ時に購入ボタンが表示されてしまうかと存じます。
"購入画面に進む"を表示する前にも売り切れているかどうかを確認するか、
売り切れている場合は何も表示されないように条件分岐を修正しましょう。

YYprogramer replied:
<% unless @item.sold_out? %>
を追記し
商品が売り切れていない場合は
購入画面に進むボタンを表示させるようにしました

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
コメントさせていただきましたので、ご確認の上、修正をお願いいたします。

修正完了後はGitHubDesktopにてコミットプッシュを行っていただき、再度Googleフォームからレビュー依頼を送信いただくようお願いいたします。

疑問点・不明点等ございましたら、コネクトメンターへお気軽にご質問ください。

YYprogramer committed:
条件分岐の追加

商品が売り切れていない場合は
購入画面に進むボタンを表示させる

YYprogramer committed:
関連付け変更に伴う修正

itemModel内でOrderModelとの関連付けをhas_manyからhas_oneに修正しました。
それに伴いOrderを複数形にしていたものを単数系の処理に変更しました

YYprogramer commented:
#what
商品購入機能の実装

#why
コードレビューの修正

□修正内容
・商品の販売状況によって、売り切れている場合は購入ボタンを表示させないように条件分岐を追加
・Itemモデル内でOrderモデルとの関連付けをhas_menyからhas_oneに修正した際、表示エラーが発生しました。orderを単数系の処理に修正。

TakashiKozuma commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def order_params
+    params.require(:order_form).permit(:number_form, :expiry_form, :cvc_form,:postal_code, :send_origin_id, :city, :street_address, :building, :phone_number, :order_id, :user_id, :item_id, :token)
```
```Rb
params.require(:order_form).permit(:number_form, :expiry_form, :cvc_form,:postal_code, :send_origin_id, :city, :street_address, :building, :phone_number, :order_id, :user_id, :item_id, :token)
```
oreder_idはストロングパラメータで送らない値なので、削除しましょう。

YYprogramer replied:
削除しました

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認致しました。
コメントを追記いたしました。
ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

YYprogramer committed:
ストロングパラメーターの修正

TakashiKozuma commented:
追加のレビュー依頼ありがとうございます！
修正点も含め、確認が完了しました。
LGTMです！！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切ってから次にお進み下さい。
